logins-sql/ffi/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+extern crate serde_json;
+extern crate rusqlite;
+extern crate logins_sql;
+extern crate sync15_adapter;
+extern crate url;
+extern crate reqwest;
+#[macro_use] extern crate log;
+
+#[cfg(target_os = "android")]
+extern crate android_logger;
+
+pub mod error;
+
+use std::os::raw::c_char;
+use std::ffi::{CString, CStr};
+
+use error::{
+    ExternError,
+    with_translated_result,
+    with_translated_value_result,
+    with_translated_void_result,
+    with_translated_string_result,
+    with_translated_opt_string_result,
+};
+
+use logins_sql::{
+    Login,
+    PasswordEngine,
+};
+
+#[inline]
+unsafe fn c_str_to_str<'a>(cstr: *const c_char) -> &'a str {
+    CStr::from_ptr(cstr).to_str().unwrap_or_default()
+}
+
+fn logging_init() {
+    #[cfg(target_os = "android")]
+    {
+        android_logger::init_once(
+            android_logger::Filter::default().with_min_level(log::Level::Trace),
+            Some("libloginsapi_ffi"));
+        debug!("Android logging should be hooked up!")
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn sync15_passwords_state_new(
+    db_path: *const c_char,
+    encryption_key: *const c_char,
+    error: *mut ExternError
+) -> *mut PasswordEngine {
+    logging_init();
+    trace!("sync15_passwords_state_new");
+    with_translated_result(error, || {
+        let path = c_str_to_str(db_path);
+        let key = c_str_to_str(encryption_key);
+        let state = PasswordEngine::new(path, Some(key))?;
+        Ok(state)
+    })
+}
+
+// indirection to help `?` figure out the target error type
+fn parse_url(url: &str) -> sync15_adapter::Result<url::Url> {
+    Ok(url::Url::parse(url)?)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn sync15_passwords_sync(
+    state: *mut PasswordEngine,
+    key_id: *const c_char,
+    access_token: *const c_char,
+    sync_key: *const c_char,
+    tokenserver_url: *const c_char,
+    error: *mut ExternError
+) {
+    trace!("sync15_passwords_sync");
```

Comment:
I think I'd rather avoid using too many macros here to keep the code from being confusing. I'll think about a way to unify the `with_translated_*`, but my concern with using a trait is that if it maps to something we don't expect then it might be extremely dangerous.